conn.go

```diff
@@ -347,6 +380,93 @@ func (conn *Conn) inWorker() {
 	}
 }
 
+// sigWorker gets signals from signal queue and handles them
+// appropriately.
+func (conn *Conn) sigWorker() {
+	for {
+		signal := conn.nextSignal()
+		if signal != nil {
+			closing := conn.handleSignal(signal)
+			if closing {
+				return
+			}
+		} else {
+			select {
+			case <-conn.signalQueueWakeUp:
+			case <-conn.shutdownSigWorker:
+				return
+			}
+		}
+	}
+}
+
+// nextSignal gets next signal from queue or nil if there was no
+// signal in queue.
+func (conn *Conn) nextSignal() *Signal {
+	conn.signalQueueLck.Lock()
+	defer conn.signalQueueLck.Unlock()
+	for len(conn.signalQueue) > 0 {
+		signal := conn.signalQueue[0]
+		conn.signalQueue = conn.signalQueue[1:]
+		if signal != nil {
+			return signal
+		}
+	}
+
+	conn.signalQueueConsumerAsleep = true
+	return nil
+}
+
+// processPendingSignalChannels either adds new pending channels to
+// signal channels list or removes them if they already existed there.
+//
+// It is assumed that conn.signalsLck is locked.
+func (conn *Conn) processPendingSignalChannels() {
+	signalsMap := make(map[chan<- *Signal]bool, len(conn.signals))
+	for _, ch := range conn.signals {
+		signalsMap[ch] = false
+	}
+	conn.pendingSignalChannelsLck.Lock()
+	for _, ch := range conn.pendingSignalChannels {
+		if _, ok := signalsMap[ch]; ok {
+			delete(signalsMap, ch)
+		} else {
+			signalsMap[ch] = false
+		}
+	}
+	conn.pendingSignalChannels = nil
+	conn.pendingSignalChannelsLck.Unlock()
+
+	conn.signals = nil
+	for ch, _ := range signalsMap {
+		conn.signals = append(conn.signals, ch)
+	}
+	signalsMap = nil
```

Comment:
But to be clear - you are asking about processPendingSignalChannels function in general or just nilling the signalsMap?

If the former then:
You can use conn.Signal to add or remove a signal channel. When you pass a channel here for the first time, then conn owns it (that means - conn is going to close this channel in conn.Close). If you pass it again, conn will release ownership, so channel is yours again. That's what actually Signal's docs are saying. Before my changes, in spite of docs, this function was just blindly adding passed channel to signal channels list. So somebody could call conn.Signal twice with the same channel and expect that conn is not going to send signals there while in reality the channel appeared twice in the signals list. So every signal could be received twice though that channel. And calling the conn.Close always panicked because of closing the closed channel.

So processPendingSignalChannels simply adds new channel to signal channels list if it wasn't there before or removes it otherwise. And I'm doing it by putting all channels into set, adding or removing them from the set and then creating new signal channels list from the set.

You could see TestRemoveSignal in signal_test.go. It is testing this case.

If the latter then well. Probably not needed.